addons/project/models/project_task_type.py

```diff
@@ -79,6 +82,44 @@ def copy(self, default=None):
             default['name'] = _("%s (copy)") % (self.name)
         return super().copy(default)
 
+    def unlink(self): # TODO: write test for deleting stages in batch (+ for multiple user in sudo)
+        # 1. All non-personal stages are processed
+        stages_to_unlink = self.filtered(lambda stage: not stage.user_id)
+
+        # 2. Personal stages are processed if the user still has at least one personal stage after unlink
+        raise_personal_stage_error = False
+        remaining_personal_stages = self.env['project.task.type'].search([('user_id', 'in', self.user_id.ids), ('id', 'not in', self.ids)], order='sequence DESC')
+        for user in self.user_id:
+            user_stages_to_unlink = self.filtered(lambda stage: stage.user_id == user)
+            user_remaining_stages = remaining_personal_stages.filtered(lambda stage: stage.user_id == user)
+            if len(user_remaining_stages) > 0:
+                stages_to_unlink |= user_stages_to_unlink
+                self._prepare_personal_stages_deletion(user_stages_to_unlink, user_remaining_stages)
```

Comment:
Even with my suggestion, it still makes one _read_group per valid iteration. Isn't there a way to improve the perf here? :slightly_smiling_face:

Reply:
> Even with my suggestion, it still makes one _read_group per valid iteration. Isn't there a way to improve the perf here? 🙂

I don't think it's a big deal as personal stages are never (as far as I know) deleted in batch for multiple users. But I could be in the future so I propose to move the records to update in the args of the method if it's ok for you.